storage: add flag to request engine type(rocksdb or pebble)

hueypark opened the pull request:
Add --test-storage-engine flag to request engine type.
Example: `make test TESTFLAGS=--test-storage-engine=rocksdb`

Release note: None

For #41620

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41742)
<!-- Reviewable:end -->

hueypark commented:
@petermattis For support `go test ./pkg/...`, there are too many unnecessary
`_ "github.com/cockroachdb/cockroach/pkg/testutils/testflag" // for test flags`.
Please let me know if there is a better alternative.

RaduBerinde reviewed (commented):
Having to add this import in test files everywhere is not great. It's not something people will remember to do when we add new tests / packages.  We need to add it in lower-level packages used by tests. For example, adding it in `serverutils/test_server_shim.go` would cover all the tests that use `serverutils` to create a test server or cluster.

Also, #41620 is specifically talking about `roachtests` (see `pkg/cmd/roachtest`), not unit tests.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41742#-:-LrdPWev4qv9qCq_ql5S:b-q1p8gs)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
Thanks for getting this started, @hueypark.

I agree with @raduberinde that adding this new import everywhere is a burden. I believe it isn't necessary. You only use `tesetflag.TestStorageEngine` in two places. Those are the only places that should need to import the flag. Is something not working that led you to add the import everywhere?

> For example, adding it in serverutils/test_server_shim.go would cover all the tests that use serverutils to create a test server or cluster.

I'd go a level deeper than `testutils/serverutils` and put logic in `server/test_server.go`. Possibly in `TestServer.Start` but I haven't looked deeply.

Another alternative is to add a new `base.EngineType`: `EngineTypeDefault`. This would be given the zero value. `serverServer.CreateEngines` could then be configured to use Pebble for the default engine based on some control mechanism. My instinct is to use an env var for this as we don't want a flag exposed in the main `cockroach` binary. The benefit of this approach is that it would provide the functionality for `roachtests` as well. We'd have to make sure this env var is propagated to `roachprod` and on to the remote commands, but that doesn't seem difficult.

> Also, #41620 is specifically talking about roachtests (see pkg/cmd/roachtest), not unit tests.

True, but doing this for unit tests is valuable too. 

@hueypark as you noticed, you also have to control the creation of temp engines. There is also `engine.NewInMem` which is used from various tests. I'd leave that alone for now (some of the `storage/engine` tests expect that to create an `*engine.RocksDB`), but I've got my eye on fixing it as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41742#-:-Lrd_reqBp55HYp4ttB2:b-e8od20)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
I think the idea was that only test binaries would have the flag (so it can't go in the `server` package)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41742#-:-LrdpM2z-OYbYXTCPDQ3:b458in9)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
@RaduBerinde @petermattis Thank you for the review.

The reason for the extensive import of `testflag` is `go test ./pkg/...` without flag result as `flag provided but not defined` and `os.Exit(2)`. There is no solution for this with the current version of Go.
- `flag provided but not defined`: https://github.com/golang/go/blob/e01e9176870d74a564cb69a70303c1c5e55c965f/src/flag/flag.go#L926
- `os.Exit(2)`: https://github.com/golang/go/blob/e01e9176870d74a564cb69a70303c1c5e55c965f/src/flag/flag.go#L979

Now I agreed that import everywhere is a burden. And reimplement with env. Could you review this again?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41742#-:-LriKMhW5DJq1Q4Smyar:b-k7yx4i)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
This is much more palatable.

Did you give consideration to my other suggestion to add `EngineTypeDefault`? The addition of `testengine.NewTempEngine` and the requirement to use that everywhere feels like something that can rot (i.e. a test can easily accidentally use `storage/engine.NewTempEngine` directly). I was imagining that `storage/engine.NewTempEngine` would be as you've changed it, but all of the tests would pass in `EngineTypeDefault`. We'd want to move `testenv.TestStorageEngine` into `storage/engine`, but I don't think those few lines of test code are problematic to include in the full binary. 

> The reason for the extensive import of testflag is go test ./pkg/... without flag result as flag provided but not defined and os.Exit(2). There is no solution for this with the current version of Go.

Huh. That's strange. I like the env var approach better anyways for this purpose.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41742#-:-LriR1o6C77cOG0ffK6y:b-je95ha)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark)

___
*[pkg/testutils/testenv/testenv.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41742#-LriPvK-3cnWcswq8lHZ:-LriPvK00NwXKVRqk0bd:bjmkxnh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6903493fdddaa5605ca1aea42ce3f9722fb13fb/pkg/testutils/testenv/testenv.go#L24)):*
> ```Go
> 
> func init() {
> 	TestStorageEngine.Set(os.Getenv("COCKROACH_TEST_STORAGE_ENGINE"))
> ```

Lint will probably complain about this. You should use `envutil.EnvOrDefaultString`.


<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
> Did you give consideration to my other suggestion to add EngineTypeDefault? The addition of testengine.NewTempEngine and the requirement to use that everywhere feels like something that can rot (i.e. a test can easily accidentally use storage/engine.NewTempEngine directly). I was imagining that storage/engine.NewTempEngine would be as you've changed it, but all of the tests would pass in EngineTypeDefault. We'd want to move testenv.TestStorageEngine into storage/engine, but I don't think those few lines of test code are problematic to include in the full binary.

I used `TestStorageEngine` instead of `EngineTypeDefault`, changing `testengine.NewTempEngine` to `engine.NewTempEngine`. And TestServer also use `TestStorageEngine`

With pebble environment some test fail and hang, but I'm not going to look further in this PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41742#-:-LrjCYhXEWAJ6UDfEYMj:b-owqdyv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis)

___
*[pkg/testutils/testenv/testenv.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41742#-LriPvK-3cnWcswq8lHZ:-LriYT8m3vHGpjU6bVhy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6903493fdddaa5605ca1aea42ce3f9722fb13fb/pkg/testutils/testenv/testenv.go#L24)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Lint will probably complain about this. You should use `envutil.EnvOrDefaultString`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thanks for the PR!

> With pebble environment some test fail and hang, but I'm not going to look further in this PR.

Ack. We probably have bugs either in Pebble or the Pebble integration. Unnecessary to fix them in this PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41742#-:-LrjRnoj1yhkEn54l88m:b7hpfrz)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

hueypark committed:
storage: add env to request engine type(rocksdb or pebble)

Add COCKROACH_TEST_STORAGE_ENGINE env to request engine type.

Release note: None

hueypark reviewed (commented):
Update commit message: `testutils:` to `storage:`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41742#-:-Lrn8oPl8WvWxN7k5Y7M:b-aq22f3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

petermattis commented:
Thanks, @hueypark. I'll kick off bors to merge this.

bors r+
